tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for lowering memref.tensor_load ops that are
+// inserted during `mhlo-legalize-to-lmhlo`.
+
+#include "mlir-hlo/Dialect/mhlo/transforms/passes.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // TF:llvm-project
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace lmhlo {
+namespace {
+
+using shape::ShapeOfOp;
+using tensor::ExtractOp;
+
+struct LegalizeTensorLoadOpPass
+    : public mlir::PassWrapper<LegalizeTensorLoadOpPass, FunctionPass> {
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<shape::ShapeDialect, tensor::TensorDialect,
+                    memref::MemRefDialect>();
+  }
+  // Perform the lowering to remove memref.tensor_load ops inserted during
+  // `mhlo-legalize-to-lmhlo`.
+  void runOnFunction() override;
+};
+
+// Converting:
+//   memref (operand) -> memref.tensor_load -> shape.shape_of
+//     to
+//   memref (operand) -> shape.shape_of
+struct ForwardShapeOfOp : public OpRewritePattern<ShapeOfOp> {
+  using OpRewritePattern<ShapeOfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ShapeOfOp shape_of,
+                                PatternRewriter& rewriter) const override {
+    auto tensor_load = shape_of.arg().getDefiningOp<memref::TensorLoadOp>();
+    if (!tensor_load) return failure();
+
+    rewriter.replaceOpWithNewOp<ShapeOfOp>(shape_of, shape_of.getType(),
+                                           tensor_load.memref());
+    return success();
+  }
+};
```

Comment:
Thanks! These patterns are relative specific to hlo-to-lhlo conversion pass. I am not sure if there are other places needing these patterns. I think we can leave it here for now and upstream them if necessary in the future (I'd like to help).

Reply:
Can you elaborate what's specific to HLO here?

Reply:
Interface `InferReturnTypeShapes` of mhlo ops are implemented on tensor level. During mhlo-to-lmhlo conversion,  the operands of a hlo op are legalized to memref type first before the op is legalized,  making it invalid to call `InferReturnTypeShapes` directly.  To solve this problem, we need to insert tensor_load op to make sure the operands passed in to the interface are valid.

This pr tries to remove these tensor_load ops after conversion in some cases. I'm not sure if these patterns are likely to be used in other places since it's highly related to the implementation of tensor-to-buffer conversion. However I agree that the implementation of these patterns has nothing to do with hlo.